Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp

```diff
@@ -626,37 +666,184 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
+
+    return cookies;
+}
+
+std::pair<String, bool> NetworkStorageSession::lookupCookiesHeaders(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, ForHTTPHeader forHTTPHeader, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, IncludeSecureCookies includeSecureCookies) const
+{
+    bool didAccessSecureCookies = false;
+    auto cookies = lookupCookies(
+        firstParty,
+        sameSiteInfo,
+        url,
+        forHTTPHeader,
+        frameID,
+        pageID,
+        applyTrackingPrevention,
+        shouldRelaxThirdPartyCookieBlocking,
+        includeSecureCookies,
+        &didAccessSecureCookies);
+
+    if (!cookies || !*cookies)
         return { { }, false };
 
-    GUniquePtr<char> cookieHeader(soup_cookies_to_cookie_header(cookies));
-    soup_cookies_free(cookies);
+    GUniquePtr<char> cookieHeader(soup_cookies_to_cookie_header(*cookies));
+    soup_cookies_free(*cookies);
 
     return { String::fromUTF8(cookieHeader.get()), didAccessSecureCookies };
 }
 
-std::pair<String, bool> NetworkStorageSession::cookiesForDOM(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking relaxThirdPartyCookieBlocking) const
+bool NetworkStorageSession::getRawCookies(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, Vector<Cookie>& rawCookies) const
+{
+    rawCookies.clear();
+
+    auto soupCookies = lookupCookies(
+        firstParty,
+        sameSiteInfo,
+        url,
+        ForHTTPHeader::No,
+        frameID,
+        pageID,
+        applyTrackingPrevention,
+        shouldRelaxThirdPartyCookieBlocking,
+        IncludeSecureCookies::Yes);
+
+    if (!soupCookies)
+        return false;
+
+    for (GSList* iter = *soupCookies; iter; iter = g_slist_next(iter)) {
+        GUniquePtr<SoupCookie> cookie(static_cast<SoupCookie*>(iter->data));
+        rawCookies.append(Cookie(cookie.get()));
+    }
+
+    g_slist_free(*soupCookies);
+
+    return true;
+}
+
+Vector<Cookie> NetworkStorageSession::domCookiesForHost(const URL& url)
 {
-    return cookiesForSession(*this, firstParty, url, sameSiteInfo, frameID, pageID, false, includeSecureCookies, applyTrackingPrevention, relaxThirdPartyCookieBlocking);
+    auto host = url.host().utf8().data();
+
+    // FIXME: libsoup needs an API so we can avoid going through all cookies.
+    GUniquePtr<GSList> soupCookies(soup_cookie_jar_all_cookies(cookieStorage()));
```

Comment:
...and I see how we have specialised this for `_free` and not `full`, so again we need to wrap elements in unique pointers :+1:

Reply:
Same here this uses `GUniquePtr<SoupCookie> soupCookie(static_cast<SoupCookie*>(iter->data));` below.